ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerBlockProductionManagerImpl.java

```diff
@@ -72,35 +67,42 @@ public ExecutionPayloadResult initiateBlockProduction(
       final boolean attemptBuilderFlow,
       final Optional<UInt64> requestedBuilderBoostFactor,
       final BlockProductionPerformance blockProductionPerformance) {
-    final ExecutionPayloadResult result;
-    if (attemptBuilderFlow) {
-      result =
-          executeBuilderFlow(
-              context, blockSlotState, requestedBuilderBoostFactor, blockProductionPerformance);
-    } else {
-      result = executeLocalFlow(context, blockSlotState, blockProductionPerformance);
-    }
-    executionResultCache.put(blockSlotState.getSlot(), result);
-    return result;
+    return executionResultCache.computeIfAbsent(
+        blockSlotState.getSlot(),
+        __ -> {
+          if (attemptBuilderFlow) {
+            return executeBuilderFlow(
+                context, blockSlotState, requestedBuilderBoostFactor, blockProductionPerformance);
+          } else {
+            return executeLocalFlow(context, blockSlotState, blockProductionPerformance);
+          }
+        });
+  }
```

Comment:
this is not guaranteed to be computed once. I think we need to synchronize.

Reply:
Modified PR, so no longer applicable